src/generate_features.py

```diff
@@ -1,3 +1,215 @@
 logger = logging.getLogger(__name__)
+logger.setLevel(logging.DEBUG)
+
+ROOT = os.getcwd()
+logger.debug('Project Root = %s', ROOT)
+
+with open(os.path.join(ROOT, 'config', 'features_config.yml'), 'r') as f:
+    config = yaml.load(f)
+with open(os.path.join(ROOT, 'data', 'external', 'cats.yml'), 'r') as f:
+    cat_map = yaml.load(f)
+    cats = list(cat_map.keys())
+logger.debug('%d categories loaded: %s', len(cats), cats)
+
+if config.get('read_from') == 'local':
+    data_dir = os.path.join(ROOT, 'data', 'external')
+elif config.get('read_from') == 's3':
+    data_dir = 's3://{}/data/external'.format(config.get('s3_bucket-name'))
+
+
+def load_data(data_dir, file_name):
+    path = os.path.join(data_dir, file_name)
+    logger.debug('Reading data from %s', path)
+    return pd.read_csv(path)
+
+
+def clean(X):
+    nans = np.isnan(X).any(axis=1)
+    logger.debug('Removing %d rows with nan values in order features (X)',
+                 sum(nans))
+    return X[~nans]
+
+
+def gmm_clust(X, n_components, cov_type):
+    logger.info('Clustering GMM with n_components=%d and cov_type=%s',
+                n_components, cov_type)
+    gmm = GMM(n_components=n_components, covariance_type=cov_type)
+    gmm.fit(X)
+    labels = gmm.predict(X)
+    return labels
+
+
+def km_clust(X, k):
+    logger.debug('Clustering KMeans with k=%d', k)
+    km = KMeans(X, k)
+    km.fit(X)
+    labels = km.labels_
+    return labels
+
+
+def gen_plots(order_types):
+    # TODO: implement heatmap plotting and save imgs
+    pass
+
+
+if __name__ == '__main__':
+    order_products = load_data(data_dir, 'order_products__train.csv')
+    order_products_prior = load_data(data_dir, 'order_products__prior.csv')
+    orders = load_data(data_dir, 'orders.csv')
+    products = load_data(data_dir, 'products.csv')
+    aisles = load_data(data_dir, 'aisles.csv')
+    departments = load_data(data_dir, 'departments.csv')
+
+    products.set_index('product_id', inplace=True)
+    orders.set_index('order_id', inplace=True)
+    aisles.set_index('aisle_id', inplace=True)
+    departments.set_index('department_id', inplace=True)
+
+    order_products = pd.concat([order_products, order_products_prior])
+    del order_products_prior
+
+    n_orders = order_products.groupby('product_id').size().rename('n_orders')
+    products = products.join(n_orders)
+
+    logger.debug('Engineering product features')
+    # Product considered "popular" if among the top_n products ordered in its department
+    top_n = config.get('popular_threshold', 5)
+
+    top_products = products.groupby('department_id')['n_orders']\
+                   .nlargest(top_n).reset_index()
+    top_products = top_products.join(products[['product_name']], on='product_id')
+    top_products = top_products.values.reshape(-1)
+    products['popular'] = products['product_name'].isin(top_products).astype(int)
+
+    products['organic'] = products['product_name'].str.lower()\
```

Comment:
not quite clear where organic fits in, why weren't other departments considered too

Reply:
organic is not related to department but instead may describe a type of food in any department and may be related to a user's behavior/preferences. Open to other suggestions on macro-level attributes that would be behavior-indicative